### DISCUSSION TITLE
Changing the CDT Build Console

### DISCUSSION THREAD

**Discussion by jamesfowkes** [eclipse-cdt/cdt#545] [Q&A]
We have an embedded C makefile project in Eclipse, which builds just fine, but now we've reached the stage of building it with CI jobs on Linux machines and docker instances, and we have added a lot of things that don't exist under Windows (e.g. `cat`, `sed` etc.)

Rather than modifying the makefile to workround this on Windows, I was hoping to make Eclipse use an WSL console instead of the default one. I have WSL installed on my laptop already.

Is there a way to tell Eclipse to use a different console for the CDT Build Console?

I have tried changing the default build command from `{cross_make}` to `wsl make`

![image](https://github.com/eclipse-cdt/cdt/assets/1223959/eda4c9bb-272f-460e-8465-f48c5b16ad15)

But this does not work because `wsl make` is interpreted as a single token:

```
"wsl make" -j16 -C ../ all 
Cannot run program "wsl make": Launching failed

Error: Program "wsl make" not found in PATH
```

**Comment by jonahgraham** [comment]
The reason this is failing is that the _Build command_ needs to be a single command, in this case `wsl` and then `make` is an argument that you add in the _Behavior_ tab.

---

That said, you are going to run into additional issues (most likely) because other parts of CDT will expect to find includes on the host (Windows) machine too.

A better bet to achieve what you are after is to use Docker tooling to build in the docker container so all your build tools are in a docker image and you can share that image among your team.

That way when you build it will run it automatically in the docker container like this:

![image](https://github.com/eclipse-cdt/cdt/assets/679236/417f6ef6-4831-4e89-a2e2-e4a0dd3f221c)

and the tooling will take care of copying files back and forth to the docker container as needed, you can see this by ctrl-click on the `#include` file and see that the path is where it copied it from on the docker container to the local machine:

![image](https://github.com/eclipse-cdt/cdt/assets/679236/50c50e02-67a5-479a-8182-c4ce08e6cc20)

To enable this, once you have docker setup, you can tell the project to build in the container in the build settings:

![image](https://github.com/eclipse-cdt/cdt/assets/679236/827d1e1f-23bc-4b6f-92e5-7982f5578def)

HTH!

**Comment by jamesfowkes** [comment]
I'm not very familiar with Docker. Is that a Docker image run from inside Eclipse somehow, or do I need to run a Docker instance completely separately to Eclipse?

I don't have anything in the "Connection" dropdown so I guess I need to install/run something, just not sure exactly what.

![image](https://github.com/eclipse-cdt/cdt/assets/1223959/5d1fb7e2-eb6a-4697-80d3-7d4f7d619919)

**Reply by jonahgraham** [reply]
Yes, you need to have docker setup already. Installing docker itself is out of scope of Eclipse CDT, but CDT will connect to existing docker. There should be lots of resources on the web, but the main docs are a good place to start https://docs.docker.com/desktop/

(Assuming you are using Ubuntu for WSL and your docker images) The same way you install tools in WSL (using apt/apt-get) is how you install tools on the docker container (using the Dockerfile)

HTH

**Comment by jamesfowkes** [comment]
OK, so I have a Docker container running and it appears in the Eclipse Docker Explorer:

![image](https://github.com/eclipse-cdt/cdt/assets/1223959/01ce6eb4-9711-4527-a39f-448476be93d6)

And I have configured it in the  project properties:

![image](https://github.com/eclipse-cdt/cdt/assets/1223959/e859a609-a95f-4eba-b7bc-0adcfffcf1c3)

But when I try and build, I get this error in the CDT Build Console:
```

18:35:30 **** Incremental Build of configuration Default for project ancillary ****
Unable to find Docker connection:hiberglobal/gcc-arm-none-eabi:latest

18:35:30 Build Finished. 0 errors, 0 warnings. (took 8ms)

```

I am stuck for what to do next, searching for that error does not return anything very useful.
